feat(csv-import): add utilities to parse duration, time, and cue numbers

yhunko-rolique opened the pull request:
Introduce `parseDurationToMs`, `parseTimeToDate`, and `parseCueNumbersColumn` for handling CSV data. Add tests for cue number parsing. Update to use `vitest` for testing.

yhunko-rolique committed:
feat(csv-import): add utilities to parse duration, time, and cue numbers

Introduce `parseDurationToMs`, `parseTimeToDate`, and `parseCueNumbersColumn` for handling CSV data. Add tests for cue number parsing. Update to use `vitest` for testing.

yhunko-rolique committed:
chore(test): refactor tests to use vitest & TS

yhunko-rolique committed:
test(moveCues): add type assertion for error in group header move test

lhermann committed:
chore: run ci tests without color

lhermann committed:
chore: update dependencies

lhermann committed:
chore: use tsup instead of native tsc to avoid JS/ESM issues

lhermann committed:
refactor: minor code formatting

lhermann committed:
chore: avoid double github action execution

yhunko-rolique committed:
refactor(csv-import): standardize file naming to use camelCase

Update file imports, exports, and references in tests to adopt camelCase naming convention. Adjust `tsconfig` to include `tests` folder.

yhunko-rolique committed:
chore(eslint): add lint script and update ESLint configuration

Add `lint` script in `package.json` for automated fixes. Migrate `eslint.config.js` to TypeScript and enhance rules. Update `tsconfig.json` to include ESLint configuration.

lhermann committed:
test: mark flaky performance test as optional

lhermann committed:
test: fix log output for markdown

lhermann commented on tests/createTimestamps.test.ts:
The createTimestamps test is one of the most important tests of the whole application. It seems like you deleted most of its test cases. This seems to have been an error.

yhunko-rolique replied:
Yeah, my bad, forgot the todo item... 🫠

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/ae54954bbe3e9cfd94aba3349b08f8fb562d74ea

There are some TS errors, but I don't think they are crucial for tests

lhermann commented on eslint.config.ts:
```diff
@@ -51,12 +55,13 @@ export default [
+    files: ['tests/**/*.ts'],
     rules: {
       '@typescript-eslint/no-unused-expressions': 'off',
+      '@typescript-eslint/no-explicit-any': 'off',
```
What is the reason for turning this off?

yhunko-rolique replied:
It was for tests. We usually need to assert something, and might do like: as any / unknown as <SpecificType>

lhermann replied:
👍

lhermann commented on tests/getChildrenTimestamps.test.ts:
```diff
@@ -63,7 +66,7 @@ describe('getChildrenTimestamps', () => {
       id: 'parent4' // Testing with undefined children
     }
-  ];
+  ] as unknown as  RundownCueOrderItem[];
```
`as RundownCueOrderItem[]` should be enough here, right?

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/477f7388da4771e5d3db5f4396a9cfaeacdca26f

lhermann commented on tests/getChildrenTimestamps.test.ts:
```diff
@@ -107,9 +110,9 @@ describe('getChildrenTimestamps', () => {
+      cues: mockTimestamps.cues
     };
-    delete incompleteTimestamps.cues['child2'];
+    delete (incompleteTimestamps.cues as never)['child2'];
```
Can you explain the `as never` usage here?

yhunko-rolique replied:
`TS2790: The operand of a delete operator must be optional.`

So in this test, we basically assert that something can be deleted, but it TS type it is stated to be required. That's why TS doesn't allow to `delete` this property.

lhermann commented on src/csv-import/parseCueNumbersColumn.ts:
```diff
@@ -0,0 +1,48 @@
+function parseCell (cell: unknown): string {
```
```suggestion
function sanitizeCellValue (cell: unknown): string {
```

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/82582929cbc7e473b4fa511daef1d33a61bfd969

lhermann commented on src/csv-import/parseCueNumbersColumn.ts:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Normalizes the cue number based on the previous cue number and current cell value.
+ */
+function normalizeCueNumber (cell: unknown, previousCueNumber?: string) {
```
Why is the `cell` type `unknown`? We do generally expect a string here right? Doesn't CSV always parse as strings? Even empty strings? What other possible value would we encounter?

yhunko-rolique replied:
it's more like a general fallback option - to use unknown when you're not sure what we can expect as a type. Papaparse library can parse cells into numbers automatically, for example.

lhermann replied:
👍

lhermann commented on src/csv-import/parseCueNumbersColumn.ts:
```diff
@@ -0,0 +1,48 @@
+
+  const [prevCue, prevGroup] = previousCueNumber.split('.')
+
+  if (cellGroup) {
```
I'd prefer `if (cellValue.includes('.'))` because `cellGroup` is not used in any other way except to see if a group separator was present.

yhunko-rolique replied:
Not only for that. I use: `parseInt(prevGroup) + 1` to add up number to the previous index inside a group. I don't believe `if (cellValue.includes('.'))` will work 🤔

lhermann replied:
Ah yes, then leave as is and it should be resolved with added documentation.

lhermann commented on src/csv-import/parseCueNumbersColumn.ts:
```diff
@@ -0,0 +1,48 @@
+
+/**
+ * Parses and normalizes a column of cue numbers.
+ */
```
Add explanation how this parsing behaves.
That cues with group separator `.` (e.g. `1.4`) always get added to the previous cue's group regardless of the parsed numbering. If no group separator is present, then they are given the next full number. The first number is `1`.

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/82582929cbc7e473b4fa511daef1d33a61bfd969

lhermann commented on src/csv-import/parseDurationToMs.ts:
```diff
@@ -0,0 +1,99 @@
+import { addSeconds, addMinutes, addHours, differenceInMilliseconds } from 'date-fns'
+
+export function parseDurationToMs (input?: unknown) {
```
1. Please define return types and briefly explain what kind of strings this function can parse.
2. I'm not quite happy with the `unknown` type. Can we really not guarantee that `input` is a string?

yhunko-rolique replied:
1. Will do
2. I was keeping bulletproof approach in mind. Unknown basically checks for the type we need / support, and handles any possible input value. But if you're really against it by all means - I can adjust the function to use `string` input.

lhermann replied:
I just wanted to double check if you have a good reason for using `unknown`. You can leave it as is.

lhermann commented on src/csv-import/parseDurationToMs.ts:
```diff
@@ -0,0 +1,99 @@
+    // Pattern 1: Colon-separated formats (HH:MM:SS, MM:SS)
+    const colonMatch = normalized.match(/^(\d{1,2}):(\d{1,2})(?::(\d{1,2}))?$/)
+    if (colonMatch) {
+      const [, first, second, third] = colonMatch.map((x) => x ? parseInt(x, 10) : undefined)
```
Why not `[, hours, minutes, seconds]`?

yhunko-rolique replied:
Probably was confused at this moment 🫠 Good catch, thanks!

lhermann commented on src/csv-import/parseDurationToMs.ts:
```diff
@@ -0,0 +1,99 @@
+
+  try {
+    // Helper to convert time components to milliseconds using date-fns
+    const toMilliseconds = (hours = 0, minutes = 0, seconds = 0) => {
```
I'll leave this reference here with a wink 😉
https://github.com/rundown-studio/package-timeutils/blob/main/src/hmsToMilliseconds.ts

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/5c67dbfc9297a8288ac67391a22f2994fe9cef48

yhunko-rolique replied:
I still need to figure why tests fail, and what's wrong with the `hmsToMilliseconds` converting logic 🤔

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+  return isValid(result) ? result : undefined
+}
+
+export function parseTimeToDate (input?: unknown, referenceDate?: Date): Date | undefined {
```
1. Please briefly explain what kind of strings this function can parse.
2. I would expect to pass a `timezone` to this function. As is this function would parse everything in the context timezone (browser default or `UTC` on server)
3. I'm not quite happy with the `unknown` type (again). Can we really not guarantee that input is a string?

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/1bb56d0ac0c0f3e72c390b93717537e0d70b4807

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+  const today = referenceDate
+    ? new Date(referenceDate.getFullYear(), referenceDate.getMonth(), referenceDate.getDate())
+    : new Date()
+  today.setHours(0, 0, 0, 0)
```
We have a utility for this that can even handle timezones: https://github.com/rundown-studio/package-timeutils/blob/main/src/getStartOfDay.ts

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/5530f98f587975fb5bd20131c0f910baf7b9d438

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+
+  try {
+    // Pattern 1: YYYY-MM-DD HH:MM[:SS] [AM/PM]
+    const isoMatch = normalized.match(/^(\d{4})-(\d{1,2})-(\d{1,2})\s+(\d{1,2}):(\d{1,2})(?::(\d{1,2}))?(?:\s*(AM|PM))?$/i)
```
I suggest we start by trying for proper ISO strings with date-fns: https://date-fns.org/v4.1.0/docs/parseISO

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/15ddd253c983f03ac8d30784026545578d5984fc

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+      }
+    }
+
+    return undefined
```
As last resort we should just try the native Date parser:
```js
  const date = Date.parse(input)
  if (!isNaN(date)) return date
```

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/fa589b743c33c61e2dcc5a1301b0b18dea07a9bd

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+// Helper function to create a date with validation
+function createValidatedDate (year: number, month: number, day: number, hour: number, minute: number, second: number): Date | undefined {
+  if (!isValidDate(year, month, day)) return undefined
+  if (!isValidTime(hour, minute, second)) return undefined
```
Can we not omit `isValidDate` and `isValidTime`. If we create the Date object and then check it's validity like in like 37 and 38, shouldn't that be enough? What edge cases are we covering here?

yhunko-rolique replied:
JS's Date constructor auto corrects invalid values automatically. That's why we need additional checks. We might get valid date object, but not valid date we intended to have.

For example: `new Date(2024, 1, 31) // March 2, 2024 (because February 31st is not valid, so it rolls over)`

Keeping these checks ensures we only accept truly valid date and time inputs.

lhermann replied:
Got it

lhermann commented on src/csv-import/parseTimeToDate.ts:
```diff
@@ -0,0 +1,213 @@
+}
+
+// Helper function to set time on today's date
+function setTimeOnDate (baseDate: Date, hour: number, minute: number, second: number): Date | undefined {
```
We already have a utility for exactly this that can also handle timezones: https://github.com/rundown-studio/package-timeutils/blob/main/src/applyDate.ts

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/5530f98f587975fb5bd20131c0f910baf7b9d438

lhermann reviewed (changes requested):
Great work, especially with the tests!
Left a few comments.

yhunko-rolique committed:
fix: https://github.com/rundown-studio/package-utils/pull/6#discussion_r2189222683

yhunko-rolique committed:
refactor(csv-import): improve cue number parsing and add detailed documentation

Rename `parseCell` to `sanitizeCellValue` for clarity. Enhance inline documentation and function descriptions for better maintainability.

yhunko-rolique committed:
refactor(csv-import): improve `parseDurationToMs` type safety and variable naming

Ensure return type is explicitly declared. Rename variables for better clarity when parsing colon-separated formats. Simplify conditional logic.

yhunko-rolique committed:
docs(csv-import): add detailed documentation for `parseDurationToMs`

Expand function documentation with examples and usage details. Include JSDoc comments for helper methods to improve maintainability and developer experience.

yhunko-rolique committed:
refactor(csv-import): replace custom duration parsing with `hmsToMilliseconds`

Simplify `parseDurationToMs` by leveraging `hmsToMilliseconds` from `@rundown-studio/timeutils`. Remove redundant helper method and update all relevant parsing logic for consistency.

yhunko-rolique committed:
fix: https://github.com/rundown-studio/package-utils/pull/6#discussion_r2189514541

yhunko-rolique committed:
test(parseTimeToDate): expand test coverage for ISO 8601 and 24-hour time formats

Add new test cases for ISO 8601 string parsing with `parseISO` and support for 24-hour time format. Adjust existing tests for updated date handling behavior.

yhunko-rolique committed:
feat(parseTimeToDate): add timezone support and enhance test coverage

Introduce optional timezone parameter for `parseTimeToDate` to handle various timezones accurately. Update implementation to leverage `@rundown-studio/timeutils` utilities. Expand test cases for different time formats, timezone scenarios, and validation logic.

yhunko-rolique committed:
test(parseTimeToDate): add setup and cleanup for timezone in tests

Introduce `beforeEach` and `afterEach` hooks to set and restore environment timezone for consistent test behavior. Update tests to exclude timezone parameter for simplified `parseTimeToDate` usage. Adjust expectations to align with updated behavior.

yhunko-rolique committed:
refactor(parseTimeToDate): update function to use options object and adjust tests

Refactor `parseTimeToDate` to accept an options object, improving parameter clarity. Update tests to align with the new signature and ensure consistent usage.

yhunko-rolique committed:
docs(parseTimeToDate): add JSDoc with examples and validation notes

Enhance `parseTimeToDate` documentation with detailed JSDoc, covering supported formats, usage examples, and validation behavior. Improve clarity and developer experience.

yhunko-rolique committed:
refactor(tests): adds missing test; refactor to Vitest

lhermann commented on tests/parseTimeToDate.test.ts:
```diff
@@ -0,0 +1,316 @@
+    // Save original TZ value
+    originalTZ = process.env.TZ
+    // Set default timezone for tests
+    process.env.TZ = 'UTC'
```
Please don't do this. The test are SPECIFICALLY set up to run in different time zones.
See `.github/workflows/unit-tests.yml`:
```
    strategy:
      fail-fast: false
      matrix:
        timezone: [UTC, America/Los_Angeles, Europe/Berlin, Australia/Sydney]
```

But, I know the problem you're trying to solve. Running different time zones means you can no longer assert against ISO strings. Instead, for tests where you DON'T PASS A TIMEZONE, you can write assertions like so:

```
    const date = parseTimeToDate('20:30')
    expect(date instanceof Date).toEqual(true)
    expect(date.getHours()).toEqual(20)
    expect(date.getMinutes()).toEqual(30)
    expect(date.getSeconds()).toEqual(0)
```

But for tests where you pass a timezone, the result should be consistent and you can assert against an ISO string.

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/1af94b6aa1048d95089262c15882f5610dd22f11

lhermann commented on tests/createTimestamps.test.ts:
```diff
@@ -35,8 +42,8 @@ const defaultCues = [
   },
-]
-const defaultCueOrder = [{ id: '#1' }, { id: '#2' }, { id: '#3' }]
+] as unknown as RundownCue[]
```
```suggestion
] as RundownCue[]
```
Or is the `unknown` crucial?

yhunko-rolique replied:
https://github.com/rundown-studio/package-utils/pull/6/commits/cf9238bcffadd22a856f2f8559cd06b05f0b85b4

lhermann commented on tests/createTimestamps.test.ts:
```diff
@@ -1111,21 +1118,21 @@ describe('createTimestamps', () => {
         ...getCueDefaults(),
         id: '#4',
-        type: 'cue',
+        type: CueType.CUE,
```
In the beginning you changed it back to `type: 'cue'` and here you change it to `type: CueType.CUE`. Is there a specific reason?

yhunko-rolique replied:
Weird TS error: `string` is not assignable to `CueType` 🤷‍♂️

lhermann replied:
🤷‍♂️

yhunko-rolique committed:
test(parseTimeToDate): simplify timezone setup and improve assertions

- Remove `beforeEach` and `afterEach` hooks for timezone setup
- Default timezone to UTC if not set in the environment
- Refactor assertions to validate key components (minutes, hours, etc.)
- Enhance consistency across test cases with more granular checks

yhunko-rolique committed:
test(createTimestamps): remove unnecessary type assertion in mock data
